ArduCopter/mode.h

```diff
@@ -1408,6 +1448,109 @@ class ModeRTL : public Mode {
 };
 
 
+#if MODE_SHIP_OPS_ENABLED == ENABLED
+class ModeShipOperation : public Mode {
+
+public:
+    // inherit constructor
+    using Mode::Mode;
+    ModeShipOperation(void);
+    Number mode_number() const override { return Number::SHIP_OPS; }
+
+    bool init(bool ignore_checks) override;
+    void run() override;
+
+    bool requires_GPS() const override { return true; }
+    bool has_manual_throttle() const override { return false; }
+    bool allows_arming(AP_Arming::Method method) const override { return true; };
+    bool is_autopilot() const override { return true; }
+    bool has_user_takeoff(bool must_navigate) const override { return true; }
+
+    bool requires_terrain_failsafe() const override { return false; }
+
+    // for reporting to GCS
+    bool get_wp(Location &loc) const override;
+
+    // SHIP_OPS states
+    enum class SubMode : uint8_t {
+        CLIMB_TO_RTL,
+        RETURN_TO_PERCH,
+        PERCH,
+        OVER_SPOT,
+        LAUNCH_RECOVERY,
+        PAYLOAD_PLACE
+    };
+    SubMode state() { return _state; }
+    const char *state_name(SubMode mode) const;
+    void set_state(SubMode mode);
+
+    // SHIP_OPS states
+    enum class KeepOutZoneMode : uint8_t {
+        NO_ACTION,
+        AVOID_KOZ,
+        EXIT_KOZ
+    };
+
+    void set_keep_out_zone_mode(KeepOutZoneMode keep_out_zone_mode);
+    KeepOutZoneMode keep_out_zone_mode() { return _keep_out_zone_mode; }
+
+    // SHIP_OPS states
+    enum class ApproachMode : uint8_t {
+        LAUNCH_RECOVERY,
+        PAYLOAD_PLACE
+    };
+
+    void set_approach_mode(ApproachMode approach_mode);
+    ApproachMode approach_mode() { return _approach_mode; }
+
+    virtual bool is_landing() const override;
+
+    // var_info for holding Parameter information
+    static const struct AP_Param::GroupInfo var_info[];
+    
+protected:
+
+    const char *name() const override { return "SHIP_OPS"; }
+    const char *name4() const override { return "SHIP"; }
+
+private:
+
+    SubMode _state = SubMode::CLIMB_TO_RTL;  // records state of rtl (initial climb, returning home, etc)
+    KeepOutZoneMode _keep_out_zone_mode = KeepOutZoneMode::NO_ACTION;  // records state of rtl (initial climb, returning home, etc)
+    ApproachMode _approach_mode = ApproachMode::LAUNCH_RECOVERY;  // records state of rtl (initial climb, returning home, etc)
+    uint32_t wp_distance() const override;
+    int32_t wp_bearing() const override;
+
+    uint32_t last_log_ms;   // system time of last time desired velocity was logging
+    float target_climb_rate;   // system time of last time desired velocity was logging
+    Vector3f offset;
+    Vector3p ship_pos_ned;
+    Vector3f ship_vel_ned;
+    Vector3f ship_accel_ned;
+    float ship_heading;
+    float ship_heading_rate;
+    float ship_heading_accel;
+    bool ship_takeoff;
+    bool pilot_correction_active;
+    bool ship_available;
```

Comment:
```suggestion
    struct {
        Vector3p pos_ned;
        Vector3f vel_ned;
        Vector3f accel_ned;
        float heading;
        float heading_rate;
        float heading_accel;
        bool available;
    } ship;

    bool ship_takeoff;
    bool pilot_correction_active;
```

.... and change things from e.g. ship_heading_rate to `ship.heading_rate`

Reply:
done